### DISCUSSION TITLE
Microsoft Graph Error: Microsoft.Kiota.Http.HttpClientLibrary: Failed to execute GetChildThreadsV2Async

### DISCUSSION THREAD

**Discussion by nguyentranenvoyair** [microsoftgraph/msgraph-sdk-dotnet#2497] [Q&A]
Context: I'm trying to create an Azure Function that pulls Teams and Channels info using Graph API. As I understand it (please correct me if I'm wrong), there's no way to send one Graph API request to retrieve all channels in my organization, and I have to first fetch all Teams and then for each Team, fetch the channels.

As there are more than 28,000 Teams, the time it takes to fetch all channels by iterating through each Team at a time is slow, so I try to fetch channels from multiple Teams concurrently, as shown by the code block in my Function below.

However, as it was almost done fetching all channels (got to 28565/28625) then I received this error:

System.Private.CoreLib: Exception while executing function: GetTeamsChannels. Microsoft.Kiota.Http.HttpClientLibrary: Failed to execute GetChildThreadsV2Async.

Could someone please help me resolve this error?

``` C# 
// ...
// ...
// ...
// the code above this fetches all Teams in my org

log.LogInformation($"Getting all channels.");

// create a semaphore to limit the number of concurrent requests
var maxConcurrentRequests = 15; // adjust this value as needed
var semaphore = new SemaphoreSlim(maxConcurrentRequests);

// counter for the number of teams done
int teamsDone = 0;

// fetch the channels for all teams concurrently, but limit the number of concurrent requests
var tasks = teamsList.Select(async team =>
{
await semaphore.WaitAsync();

try
{
    log.LogInformation($"Fetching channels for team: {team.Id} - {team.DisplayName}");
    var channels = await graphClient.Teams[team.Id].Channels.GetAsync();

    foreach (var channel in channels.Value)
    {
        TeamsChannels.Add((team.Id, team.DisplayName, team.Mail,
                            channel.Id, channel.DisplayName, channel.Email,
                            channel.Description));
        ++channelCount;
    }

    // log.LogInformation($"Team: {team.DisplayName} has {channels.Value.Count} channels.");

    // increment the counter and log the progress
    var done = Interlocked.Increment(ref teamsDone);
    log.LogInformation($"FETCHING CHANNELS PROGRESS: {done}/{teamsList.Count}");
}
finally
{
    // Release is called to release the semaphore, allowing another request to be made
    semaphore.Release();
}
});
await Task.WhenAll(tasks); 
```

**Comment by andrueastman** [comment]
Thanks for raising this @nguyentranenvoyair 

Any chance you have looked into batching the requests using json batching? https://learn.microsoft.com/en-us/graph/sdks/batch-requests?tabs=csharp

**Reply by nguyentranenvoyair** [reply]
Hello, yes I've tried batching according to the SDK's instructions and it worked. 

However, I figured out that the error is likely caused by some invalid team ids (I don't exactly understand how this is the case because I got these team ids directly from Graph API, perhaps these teams got soft deleted or something?). I noticed this because although my function now completed successfully, there's some teams that failed to fetch channels for.

So, some team ids were invalid for some reason and as a result, that is the error returned after making requests ie.

``` C#
graphClient.Teams[teamId].Channels.GetAsync()
```


Do you know any possible reasons why?

**Reply by andrueastman** [reply]
When calling this code, you should ideally have a catch block for when the API returns an error. From the error you can obtain the error information from the API as outlined at https://github.com/microsoftgraph/msgraph-sdk-dotnet/blob/dev/docs/errors.md
```cs
var channels = await graphClient.Teams[team.Id].Channels.GetAsync();
```
